### PR TITLE
Feature: ProxyPNL

### PR DESCRIPTION
**Pull request by Magnus167**

### CODE REVIEW

File: macrosynergy/pnl/historic_portfolio_volatility.py [hunk 1]
```diff
@@ -94,33 +97,56 @@ def flat_std(x: np.ndarray, remove_zeros: bool = True) -> float:
...
-def _single_calc(
-    row,
+def _rolling_window_calc(
+    real_date_pdt: pd.Timestamp,
     ticker_df: pd.DataFrame,
     lback_periods: int,
     nan_tolerance: float,
     roll_func: callable,
     remove_zeros: bool,
     weights: Optional[np.ndarray] = None,
```
Comment (lsimonsen) [inline macrosynergy/pnl/historic_portfolio_volatility.py:107]: As far as I can see neither of these three options are currently being used?

File: macrosynergy/pnl/historic_portfolio_volatility.py [hunk 2]
```diff
@@ -11,15 +11,18 @@
...
     ticker_df_to_qdf,
+    get_eops,
+    get_cid,
 )
-from macrosynergy.panel.historic_vol import get_cycles
+
+RETURN_SERIES_XCAT = "_PNL_USD1S_ASD"
 
 
 def expo_weights(lback_periods: int = 21, half_life: int = 11):
```
Comment (lsimonsen) [inline macrosynergy/pnl/historic_portfolio_volatility.py:28]: Should all this imports not be:
```python
from macrosynergy.panel.historic_vol import expo_weights, expo_std, flat_std
```

Comment (lsimonsen) [reply macrosynergy/pnl/historic_portfolio_volatility.py:28]: Instead of specify the same functions in two different versions.

File: macrosynergy/pnl/historic_portfolio_volatility.py [hunk 3]
```diff
@@ -11,15 +11,18 @@
 sys.path.append(os.getcwd())
```
Comment (lsimonsen) [inline macrosynergy/pnl/historic_portfolio_volatility.py:11]: Sorry, I hadn't spotted this one. We should never modify the clients Current Working Directory in my books.

Any reason why @Magnus167 ?

Comment (Magnus167) [reply macrosynergy/pnl/historic_portfolio_volatility.py:11]: These are solely for the initial phase of the project to avoid having to install/re-install between each debug session, as the project has several changes between consecutive review sessions.
Rest assured, these changes will not make it to `develop` (or beyond) thanks to [check_sys_path.sh](https://github.com/macrosynergy/macrosynergy/blob/develop/.github/scripts/check_sys_path.sh) in our unittests

File: macrosynergy/pnl/historic_portfolio_volatility.py [hunk 4]
```diff
@@ -11,15 +11,18 @@
...
     standardise_dataframe,
     is_valid_iso_date,
     qdf_to_ticker_df,
     ticker_df_to_qdf,
+    get_eops,
+    get_cid,
 )
-from macrosynergy.panel.historic_vol import get_cycles
+
+RETURN_SERIES_XCAT = "_PNL_USD1S_ASD"
```
[the remark this answers is not in the document]
Comment (Magnus167) [reply macrosynergy/pnl/historic_portfolio_volatility.py:25]: Once we have a clear entry point in place, we'll make this configurable as well.

File: macrosynergy/pnl/historic_portfolio_volatility.py (hunk 1 above)
Comment (Magnus167) [reply macrosynergy/pnl/historic_portfolio_volatility.py:107]: they are passed in line 273 (declared as `_args` , updated in 246, 248 ...)

Comment (Magnus167) [reply macrosynergy/pnl/historic_portfolio_volatility.py:107]: Currently, they are not used as the univariate volatility calc. is commented out (L135->L147).

File: macrosynergy/management/simulate/simulate_quantamental_data.py [hunk 5]
```diff
@@ -366,6 +368,110 @@ def make_test_df(
...
+
+    df_rtn = pd.DataFrame(index=dates, data=rtn)
+    df_signals = pd.DataFrame(index=dates, data=signals)
+    # TODO change dates to be previous month...
+
+    # TODO stack into quantamental dataframe
+    # return df_rtn, df_signals
+    xr_tickers = [f"{cid}_{xcat}_{return_suffix}" for cid in cids]
+    csig_tickers = [f"{cid}_{xcat}_{signal_suffix}" for cid in cids]
+    dfR = pd.concat([df_rtn, df_signals], axis=1)
```
Comment (lsimonsen) [inline macrosynergy/management/simulate/simulate_quantamental_data.py:469]: Python naming convention is no camel case for variable - so `df_r` not `dfR`.

**Review by lsimonsen — APPROVED** [review]
Still working my way through it, but a few points above.

**Comment by lsimonsen** [comment]
Can we split the PR into smaller PRs such as:
* `feature/proxy_pnl/transaction_costs`
* `feature/proxy_pnl/estimate_volatility`
* `feature/proxy_pnl/contract_signals`

File: macrosynergy/management/simulate/simulate_quantamental_data.py (hunk 5 above)
Comment (lsimonsen) [reply macrosynergy/management/simulate/simulate_quantamental_data.py:482]: FYI @Magnus167 -> [PEP-0008](https://peps.python.org/pep-0008/) :eyes:

File: macrosynergy/management/utils/core.py [hunk 6]
```diff
@@ -322,3 +321,25 @@ def rec_search_dict(d: dict, key: str, match_substring: bool = False, match_type
                 return result
 
     return None
+
+
+class Timer(object):
```
Comment (lsimonsen) [inline macrosynergy/management/utils/core.py:326]: One of us need to write a unit-test for this one. Should be fairly straight forward. Any preferences/

File: macrosynergy/pnl/contract_signals.py [hunk 7]
```diff
@@ -69,17 +65,20 @@ def _check_arg_types(
...
-            df is None or len(df) > 0,
-            cids is None or len(cids) > 0,
-            ctypes is None or len(ctypes) > 0,
-            cscales is None or len(cscales) > 0,
-            csigns is None or len(csigns) > 0,
-            hbasket is None or len(hbasket) > 0,
-            hscales is None or len(hscales) > 0,
-        ]
-    )
+    try:
```
Comment (lsimonsen) [inline macrosynergy/pnl/contract_signals.py:68]: If I am honest, I think we are trying to be too clever for the data-type checking.

Personally I would prefer something like: `assert df is None or (isinstance(df, pd.DataFrame) and not df.empty)`. Although it takes up more lines, being more explicit about what we are checking for it useful.

File: macrosynergy/pnl/contract_signals.py [hunk 8]
```diff
@@ -243,11 +242,32 @@ def _apply_hedge_ratios(
...
 
+def _add_hedged_signals(
+    df_contract_signals: QuantamentalDataFrame,
+    df_hedge_signals: Optional[QuantamentalDataFrame] = None,
+) -> QuantamentalDataFrame:
+    if df_hedge_signals is None:
+        return df_contract_signals
+
+    dfC: pd.DataFrame = qdf_to_ticker_df(df_contract_signals)
+    dfH: pd.DataFrame = qdf_to_ticker_df(df_hedge_signals)
```
Comment (lsimonsen) [inline macrosynergy/pnl/contract_signals.py:261]: See PEP-8 comment above.

File: macrosynergy/pnl/historic_portfolio_volatility.py [hunk 9]
```diff
@@ -2,125 +2,74 @@
...
-    if remove_zeros:
-        x = x[x != 0]
-        w = w[0 : len(x)] / sum(w[0 : len(x)])
-    w = w / sum(w)  # weights are normalized
-    mabs = np.sum(np.multiply(w, np.abs(x)))
-    return mabs
-
-
-def flat_std(x: np.ndarray, remove_zeros: bool = True) -> float:
+def _univariate_volatility(
```
Comment (lsimonsen) [inline macrosynergy/pnl/historic_portfolio_volatility.py:41]: Are we using the `_univariate_volatility` function? If not, we should remove it.

File: tests/integration/pnl/test_historic_portfolio_volatility.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+                est_freq="m",
+                lback_periods=21,
+                lback_meth="ma",
+                cids=['XX' + str(list(signal_types[2 + i]).index(1))],
+                start=start.strftime("%Y-%m-%d"),
+                end=end.strftime("%Y-%m-%d"),
+            ).dropna().reset_index(drop=True)
+            hvol, df_pvol
+
+        results
```
Comment (lsimonsen) [inline tests/integration/pnl/test_historic_portfolio_volatility.py:168]: Why need to add the test of what value we expect - such as `(0, 0, 0)` being zero volatility for all periods.

**Review by lsimonsen — CHANGES_REQUESTED** [review]
First round of review: a few changes need, and need a deeper dive from my side.

File: macrosynergy/pnl/historic_portfolio_volatility.py [hunk 11]
```diff
@@ -129,173 +78,234 @@ def _single_calc(
...
+    x: np.ndarray, y: np.ndarray, w: np.ndarray = None, remove_zeros: bool = True
+):
+    """
+    Estimate covariance between two series after applying weights.
 
-    return univariate_vol
+    """
+    # assert len(x) == len(w), "weights and window must have same length"
+    if remove_zeros:
+        x = x[x != 0]
```
Comment (lsimonsen) [inline macrosynergy/pnl/historic_portfolio_volatility.py:96]: We need to filter out `y` as well (and check the comparison:

```python
assert (len(x) == len(y)) and (len(x) <= len(w))
if remove_zeros:
    mask = (x != 0) & (y != 0)
    x = x[mask]
    y = y[mask]
    w = w[0 : len(x)] 
else:
    assert len(x) == len(w)
w = w / sum(w)  # weights are normalized
```
